### PR TITLE
feat: streaming parser for eszip v2

### PR DESCRIPTION
**Pull request by littledivy**
Implements the streaming decoder (a tokio_util codec) and a tiny encoder for the tests (to be replaced after module graphs integration in #40)

cc @lucacasonato

### CODE REVIEW

File: examples/loader.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+          (specifier, source_ptr.0, source_ptr.1)
+        }
+        // FIXME: Maybe poll here for source here?
+        HeaderFrame::Redirect(..) => unimplemented!(),
+      };
+
+      let mut buf = buf.lock().unwrap();
+      // FIXME: Offset calculation hack.
+      // Maybe offsets should be from the beginning of the file
+      // instead of the data section?
```
Comment (lucacasonato) [inline examples/loader.rs:87]: Can't do that as then we'd have to do a double pass during eszip building: first figure out the length of the data section by building it once with fake offsets, and then do it again with the real offsets.

File: examples/loader.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+          unreachable!()
+        }
+      };
+      let code = String::from_utf8_lossy(source).to_string();
+      println!("[loader] code: {}", code);
+      Ok(ModuleSource {
+        code,
+        module_url_specified: specifier.to_string(),
+        module_url_found: new_specifier.to_string(),
+        module_type: ModuleType::JavaScript,
```
Comment (lucacasonato) [inline examples/loader.rs:111]: Should be switched based on `source.kind`.

File: examples/loader.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  let loader = StreamingLoader {
+    headers: Arc::clone(&headers),
+  };
+
+  let mut js_runtime = JsRuntime::new(RuntimeOptions {
+    module_loader: Some(Rc::new(loader)),
+    ..Default::default()
+  });
+
+  let main_module = "file://main.js/".parse().unwrap();
```
Comment (lucacasonato) [inline examples/loader.rs:142]:
```suggestion
  let main_module = "file:///main.js".parse().unwrap();
```

**Review by lucacasonato — COMMENTED** [review]
This is not going to work right: if the loader is waiting for a module to arrive, and it arrives to be a redirect, the loader will not be woken. This will cause a hang, or an error in the loader.

We are not yet notifying on Header frame insertion. I think this could be significantly simplified if we didn't try to parse the headers in a streaming fashion.

1. First wait for the entire headers to be received and parsed.
2. Then create the `StreamingLoader` and deal with the module bodies.

That way we don't need to deal with waking on header frame receive.

**Comment by littledivy** [comment]
Closing in favor of #40

**Pull request closed without merging**